pennylane/tape/tapes/reversible.py

```diff
@@ -92,16 +92,20 @@ def _matrix_elem(vec1, obs, vec2, device):
             vec1 (array[complex]): a length :math:`2^N` statevector
             obs (.Observable): a PennyLane observable
             vec2 (array[complex]): a length :math:`2^N` statevector
-            device (.QubitDevice): the device used to compute the matrix elements
+            wires (pennylane.wires.Wires): wires of the device used to prepare the state
         """
         # pylint: disable=protected-access
-        mat = device._reshape(obs.matrix, [2] * len(obs.wires) * 2)
-        wires = obs.wires
 
-        vec1_indices = ABC[: device.num_wires]
+        mat = np.reshape(obs.matrix, [2] * len(obs.wires) * 2)

```

Comment:
Yep! Backprop is so much faster!

Reply:
But that now makes sense, I was wondering why the device is used. 

One thing I was wondering too is hardware compatibility. Since we can always write a qubit circuit with PaulZ observables (by rotating the basis), and PauliZ is unitary, can't we implement reversible diff by applying the original unitary, observable, and diffed unitary inverse and measure the overlap with the zero state?

Reply:
Yes, you could apply this method in HW (one advantage that it has over other reversible adjoint methods). But I don't think you would want to do that. Pretty much any gate whose generator is also a gate also admits a parameter-shift rule. The reversible method might save you circuit evaluations, but results in deeper circuits. (the main advantage is memory savings compared to backprop on simulators)